src/governance.cpp

```diff
@@ -514,46 +494,43 @@ bool CGovernanceManager::UpdateGovernanceObject(CGovernanceVote& vote, CNode* pf
             //   otherwise we'll think a full sync succeeded when they return a result
             if(!masternodeSync.IsSynced()) return false;
 
-            LogPrintf("CGovernanceManager::UpdateGovernanceObject - Unknown proposal %d, asking for source proposal\n", vote.nParentHash.ToString());
+            LogPrintf("CGovernanceManager::UpdateGovernanceObject - Unknown object %d, asking for source\n", vote.nParentHash.ToString());
             mapOrphanVotes[vote.nParentHash] = vote;
 
-            if(!mapAskedForGovernanceObject.count(vote.nParentHash)){
-                pfrom->PushMessage(NetMsgType::MNGOVERNANCEVOTESYNC, vote.nParentHash);
-                mapAskedForGovernanceObject[vote.nParentHash] = GetTime();
-            }
+            SyncParentObjectByVote(pfrom, vote);
         }
 
-        strError = "Proposal not found!";
+        strError = "Governance object not found!";
         return false;
     }
 
 
     return AddOrUpdateVote(vote, strError);
 }
 
-bool CGovernanceManager::AddOrUpdateVote(CGovernanceVote& vote, std::string& strError)
+bool CGovernanceManager::AddOrUpdateVote(const CGovernanceVote& vote, std::string& strError)
 {
     LOCK(cs);
 
-    // store newest vote per action
-    arith_uint256 a = UintToArith256(vote.nParentHash) + vote.nVoteSignal;
-    uint256 hash2 = ArithToUint256(a);
+    // GET DETERMINISTIC HASH INCLUDING PARENT/TYPE
+    uint256 nTypeHash = vote.GetTypeHash();
+    uint256 nHash = vote.GetHash();
 
-    if(mapVotes.count(hash2))
-    {
-        if(mapVotes[hash2].nTime > vote.nTime){
-            strError = strprintf("new vote older than existing vote - %s", vote.GetHash().ToString());
+    if(mapVotesByType.count(nTypeHash)) {
+        if(mapVotesByType[nTypeHash].nTime > vote.nTime){
+            strError = strprintf("new vote older than existing vote - %s", nTypeHash.ToString());
             LogPrint("mngovernance", "CGovernanceObject::AddOrUpdateVote - %s\n", strError);
             return false;
         }
-        if(vote.nTime - mapVotes[hash2].nTime < GOVERNANCE_UPDATE_MIN){
-            strError = strprintf("time between votes is too soon - %s - %lli", vote.GetHash().ToString(), vote.nTime - mapVotes[hash2].nTime);
+        if(vote.nTime - mapVotesByType[nTypeHash].nTime < GOVERNANCE_UPDATE_MIN){
+            strError = strprintf("time between votes is too soon - %s - %lli", nTypeHash.ToString(), vote.nTime - mapVotesByType[nHash].nTime);
```

Comment:
in general: looks much cleaner now so after this ^^^ is addressed imo we are good to merge